agent/utils.mjs

```diff
@@ -0,0 +1,659 @@
+            } else if( c > 239 && c < 248 ) {
+                if( i + 2 >= arrBytes.length )
+                    throw new Error( "UTF-8 decode: incomplete 4-byte sequence" );
+                c = ( c & 7 ) <<
+                    18 | ( arrBytes[i++] & 63 ) << 12 |
+                    ( arrBytes[i++] & 63 ) << 6 | arrBytes[i++] & 63;
+            } else {
+                throw new Error(
+                    "UTF-8 decode: unknown multi-byte start 0x" +
+                    c.toString( 16 ) + " at index " + ( i - 1 )
+                );
+            }
+        }
+        if( c <= 0xffff )
+            s += String.fromCharCode( c ); else if( c <= 0x10ffff ) {
+            c -= 0x10000;
+            s += String.fromCharCode( c >> 10 | 0xd800 );
+            s += String.fromCharCode( c & 0x3FF | 0xdc00 );
+        } else {
+            throw new Error(
+                "UTF-8 decode: code point 0x" +
+                c.toString( 16 ) + " exceeds UTF-16 reach" );
+        }
+    }
+    return s;
+}
+
+export function hexToBytes( strHex, isInversiveOrder ) { // convert a hex string to a byte array
+    isInversiveOrder = !!(
+        ( isInversiveOrder != null && isInversiveOrder != undefined && isInversiveOrder )
+    );
+    strHex = strHex || "";
+    strHex = "" + strHex;
+    strHex = strHex.trim().toLowerCase();
+    if( strHex.length > 1 && strHex[0] == "0" && ( strHex[1] == "x" || strHex[1] == "X" ) )
+        strHex = strHex.substr( 2, strHex.length - 2 );
+    if( ( strHex.length & 1 ) !== 0 )
+        strHex = "0" + strHex;
+    const cnt = strHex.length;
+    let i, j;
+    const arrBytes = new Uint8Array( cnt / 2 );
+    for( i = 0, j = 0; i < cnt; ++j, i += 2 )
+        arrBytes[j] = parseInt( strHex.substr( i, 2 ), 16 );
+    if( isInversiveOrder )
+        invertArrayItemsLR( arrBytes );
+    return arrBytes;
+}
+
+export function bytesToHex( arrBytes, isInversiveOrder ) { // convert a byte array to a hex string
+    isInversiveOrder = !!(
+        ( isInversiveOrder != null && isInversiveOrder != undefined && isInversiveOrder )
+    );
+    const hex = [];
+    for( let i = 0; i < arrBytes.length; i++ ) {
+        const current = arrBytes[i] < 0 ? arrBytes[i] + 256 : arrBytes[i];
+        const c0 = ( current >>> 4 ).toString( 16 );
+        const c1 = ( current & 0xF ).toString( 16 );
+        if( isInversiveOrder ) {
+            hex.splice( 0, 0, c0 );
+            hex.splice( 1, 0, c1 );
+        } else {
+            hex.push( c0 );
+            hex.push( c1 );
+        }
+    }
+    return hex.join( "" );
+}
+
+export function bytesAlignLeftWithZeroes( arrBytes, cntMin ) {
+    const arrOneZeroByte = new Uint8Array( 1 );
+    arrOneZeroByte[0] = 0;
+    while( arrBytes.length < cntMin )
+        arrBytes = bytesConcat( arrOneZeroByte, arrBytes );
+    return arrBytes;
+}
+
+export function bytesAlignRightWithZeroes( arrBytes, cntMin ) {
```

Comment:
You can change it to ethers.utils.zeroPad

Reply:
Rewritten. Simplified. No `for`/`while` loops.